plugins/zenoh-plugin-storage-manager/src/replica/aligner.rs

```diff
@@ -134,267 +140,208 @@ impl Aligner {
-            }
+                (other_subintervals, no_err)
+            };
+            // get intervals diff
+            (this.get_subinterval_diff(other_subintervals), no_err)
+        } else {
+            (HashSet::new(), true)
         }
-        Vec::new()
     }
 
-    //perform hot alignment
-    // if HOT misaligned, ask for content(timestamps) of misaligned subintervals
-    async fn perform_hot_alignment(
+    async fn get_content_diff(
         &self,
-        this: &super::Digest,
-        other_rep: String,
-        other: super::Digest,
-    ) -> Vec<LogEntry> {
-        // get interval hashes for HOT intervals from other
-        let other_intervals = other.get_era_content(super::EraType::Hot);
-        // get era diff
-        let diff_intervals = this.get_interval_diff(other_intervals);
-
-        // get subintervals for mismatching intervals from other
-        let other_subintervals = other.get_interval_content(diff_intervals);
-        // get intervals diff
-        let diff_subintervals = this.get_subinterval_diff(other_subintervals);
-
+        diff_subintervals: HashSet<u64>,
+        this: &Digest,
+        other: &Digest,
+        other_rep: &str,
+    ) -> (Vec<LogEntry>, bool) {
         if !diff_subintervals.is_empty() {
             let mut diff_string = Vec::new();
-            // properties = timestamp=xxx,subintervals=[www,ee,rr]
             for each_sub in diff_subintervals {
                 diff_string.push(each_sub.to_string());
             }
             let properties = format!(
                 "timestamp={}&{}=[{}]",
                 other.timestamp,
-                super::SUBINTERVALS,
+                SUBINTERVALS,
                 diff_string.join(",")
             );
             // expecting sample.value to be a vec of log entries with their checksum
-            let reply_content = self.perform_query(other_rep.to_string(), properties).await;
+            let (reply_content, mut no_err) =
+                self.perform_query(other_rep.to_string(), properties).await;
             let mut other_content: HashMap<u64, Vec<LogEntry>> = HashMap::new();
             for each in reply_content {
                 match serde_json::from_str(&each.value.to_string()) {
                     Ok((i, c)) => {
                         other_content.insert(i, c);
                     }
-                    Err(e) => error!("Error decoding reply: {}", e),
+                    Err(e) => {
+                        error!("[ALIGNER] Error decoding reply: {}", e);
+                        no_err = false;
+                    }
                 };
             }
             // get subintervals diff
-            return this.get_full_content_diff(other_content);
+            let result = this.get_full_content_diff(other_content);
+            return (result, no_err);
+        }
+        (Vec::new(), true)
+    }
+
+    async fn perform_query(&self, from: String, properties: String) -> (Vec<Sample>, bool) {
```

Comment:
`from` should be a `&str` since `perform_query()` is always called passing a '<&str>.to_string()`. This will avoid String allocation.

Better: could it be a `&keyexpr` or a `&OwnedKeyExpr` ?
If yes, the `selector` below can be computed without validity/canonical check as such:
```rust
let selector = (KeyExpr::from(&self.digest_key) / from).with_parameters(&properties);
```

Reply:
`from` indicates the replica that this replica needs to talk to in order to fetch the missing updates. We can convert the type from `&str` into a `&keyexpr` or a `&OwnedKeyExpr`, but I thought it might not be intuitive.  I agree though that it would reduce canonical check inside the function that is frequently used. 
Should I proceed with making it a `&OwnedKeyExpr`?

Reply:
Let's do it in a separate PR after this one, so we better track/explain it.

Reply:
Agreed